controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,495 @@
+//   in the GenerateRequest.
+// * Eventually the patched templates are used to update the specs of the desired objects.
+func Apply(ctx context.Context, blueprint *scope.ClusterBlueprint, desired *scope.ClusterState) error {
+	log := tlog.LoggerFrom(ctx)
+
+	// Create a patch generation request.
+	req, err := request(blueprint, desired)
+	if err != nil {
+		return errors.Wrapf(err, "failed to generate patch request")
+	}
+
+	// Loop over patches in ClusterClass, generate patches and apply them to the request.
+	for i := range blueprint.ClusterClass.Spec.Patches {
+		patch := blueprint.ClusterClass.Spec.Patches[i]
+		ctx, log = log.WithValues("patch", patch.Name).Into(ctx)
+
+		log.V(5).Infof("Applying patch to templates")
+
+		// Create patch generator for the current patch.
+		generator, err := createPatchGenerator(&patch)
+		if err != nil {
+			return err
+		}
+
+		// Generate patches.
+		resp, err := generator.Generate(ctx, req)
+		if err != nil {
+			return errors.Errorf("failed to generate patches for patch %q", patch.Name)
+		}
+
+		// Apply patches to the request.
+		if err := applyPatchesToRequest(ctx, req, resp); err != nil {
+			return err
+		}
+	}
+
+	// Use patched templates to update the desired state objects.
+	log.V(5).Infof("Applying patches to desired state")
+	if err := updateDesiredState(ctx, req, blueprint, desired); err != nil {
+		return errors.Wrapf(err, "failed to apply patches to desired state")
+	}
+
+	return nil
+}
+
+// request creates a GenerateRequest based on the ClusterBlueprint and the desired state.
+// ClusterBlueprint supplies the templates. Desired state is used to calculate variables which are later used
+// as input for the patch generation.
+// NOTE: GenerateRequestTemplates are created for the templates of each individual MachineDeployment in the desired
+// state. This is necessary because some builtin variables are MachineDeployment specific. For example version and
+// replicas of a MachineDeployment.
+// NOTE: A single GenerateRequest object is used to carry templates state across subsequent Generate calls.
+func request(blueprint *scope.ClusterBlueprint, desired *scope.ClusterState) (*api.GenerateRequest, error) {
+	req := &api.GenerateRequest{}
+
+	// Calculate global variables.
+	globalVariables, err := variables.Global(blueprint.Topology, desired.Cluster)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate global variables")
+	}
+	req.Variables = globalVariables
+
+	// Add the InfrastructureClusterTemplate.
+	t, err := newTemplate(blueprint.InfrastructureClusterTemplate, api.TemplateID{
+		APIVersion: blueprint.InfrastructureClusterTemplate.GetAPIVersion(),
+		Kind:       blueprint.InfrastructureClusterTemplate.GetKind(),
+		TargetType: api.InfrastructureClusterTemplateTargetType,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create InfrastructureCluster template %s",
```

Comment:
```suggestion
		return nil, errors.Wrapf(err, "failed to prepare InfrastructureCluster template %s for patching",
```
I would avoid using create given that everything happens in memory
Same for other error message in this func